File: DailyLeetcode/11Aug/RangeProductQueriesofPowers.cpp
// Time complexity: O(q×logn)
// Space complexity: O(logn)
class Solution
{
public:
    int mod = 1e9 + 7;
    // decompose powers
    vector<int> power(int n)
    {
        int rep = 1;
        vector<int> res;
        while (n)
        {
            if (n % 2 == 1)
            {
                res.push_back(rep);
            }
            n /= 2;
            rep *= 2;
        }
        return res;
    }
    vector<int> productQueries(int n, vector<vector<int>> &queries)
    {
        // direct calculation
        vector<int> p = power(n);
        vector<int> result;
        for (auto it : queries)
        {
            int ans = 1;
            int st = it[0];
            int en = it[1];
            for (int i = st; i <= en; i++)
            {
                ans = (long long)ans * p[i] % mod;
            }
            result.push_back(ans);
        }
        return result;
    }
};

File: DailyLeetcode/11Aug/RangeProductQueriesofPowers.py
class Solution:
    def __init__(self):
        self.mod = int(1e9 + 7)

    # decompose powers
    def power(self, n):
        rep = 1
        res = []
        while n:
            if n % 2 == 1:
                res.append(rep)
            n //= 2
            rep *= 2
        return res

    def productQueries(self, n, queries):
        # direct calculation
        p = self.power(n)
        result = []
        for st, en in queries:
            ans = 1
            for i in range(st, en + 1):
                ans = (ans * p[i]) % self.mod
            result.append(ans)
        return result
